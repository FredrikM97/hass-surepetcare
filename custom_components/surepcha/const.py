""" "Constants for the Sure Petcare integration."""

DOMAIN = "surepcha"

CONF_EMAIL = "email"
CONF_PASSWORD = "password"

TOKEN = "token"
CLIENT_DEVICE_ID = "client_device_id"
FACTORY = "surepetcare_factory"
TIMEOUT_API = 10
KEY_API = "surepetcare_api"
COORDINATOR_LIST = "coordinator_list"
COORDINATOR_DICT = "coordinator_dict"
COORDINATOR = "coordinator"
ENTRY_ID = "entry_id"
SCAN_INTERVAL = 300
POLLING_SPEED = "polling_speed"
LOCATION_INSIDE = "location_inside"
LOCATION_OUTSIDE = "location_outside"
OPTION_DEVICES = "devices"
DEVICE_OPTION = "device_option"
OPTIONS_FINISHED = "finished"
PRODUCT_ID = "product_id"
DEVICES = "devices"
NAME = "name"
