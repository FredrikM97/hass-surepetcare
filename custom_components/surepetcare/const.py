"""TODO."""

DOMAIN = "surepetcare"

CONF_EMAIL = "email"
CONF_PASSWORD = "password"

TOKEN = "token"
CLIENT_DEVICE_ID = "client_device_id"
FACTORY = "surepetcare_factory"
TIMEOUT_API = 10
KEY_API = "surepetcare_api"
COORDINATOR_LIST = "coordinator_list"
COORDINATOR_DICT = "coordinator_dict"
COORDINATOR = "coordinator"
