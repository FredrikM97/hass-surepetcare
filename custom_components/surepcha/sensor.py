"""Support for Sure Petcare sensors."""

from dataclasses import dataclass
import logging
from types import MappingProxyType
from typing import Any

from surepcio import SurePetcareClient
from surepcio.enums import ProductId, PetLocation
from surepcio.devices import Pet
from surepcio.devices.pet import PetPositionResource
from homeassistant.components.sensor import (
    SensorDeviceClass,
    SensorEntity,
    SensorEntityDescription,
    SensorStateClass,
)
from homeassistant.helpers.entity import EntityCategory
from homeassistant.config_entries import ConfigEntry
from homeassistant.const import PERCENTAGE, UnitOfMass, UnitOfVolume
from homeassistant.core import HomeAssistant
from homeassistant.helpers.entity_platform import AddConfigEntryEntitiesCallback

from .const import (
    COORDINATOR,
    COORDINATOR_DICT,
    DOMAIN,
    KEY_API,
    LOCATION_INSIDE,
    LOCATION_OUTSIDE,
    NAME,
    OPTION_DEVICES,
    PRODUCT_ID,
)
from .coordinator import SurePetCareDeviceDataUpdateCoordinator
from .entity import (
    SurePetCareBaseEntity,
    SurePetCareBaseEntityDescription,
)
from .helper import MethodField, index_attr, option_name, should_add_entity, sum_attr

logger = logging.getLogger(__name__)


def get_location(
    device: Pet, entry_options: MappingProxyType[str, Any]
) -> PetLocation | str | None:
    """Return PetLocation, or None if unknown.

    Uses reconfigured values for location_inside/location_outside if available.
    """
    position: PetPositionResource = getattr(device.status, "activity", None)

    if position is not None:
        if position.where == PetLocation.INSIDE:
            return (
                entry_options[OPTION_DEVICES]
                .get(str(position.device_id), {})
                .get(LOCATION_INSIDE, position.where)
            )
        elif position.where == PetLocation.OUTSIDE:
            return (
                entry_options[OPTION_DEVICES]
                .get(str(position.device_id), {})
                .get(LOCATION_OUTSIDE, position.where)
            )

    return None


@dataclass(frozen=True, kw_only=True)
class SurePetCareSensorEntityDescription(
    SurePetCareBaseEntityDescription, SensorEntityDescription
):
    """Describes SurePetCare sensor entity."""


SENSOR_DESCRIPTIONS_BATTERY: tuple[SurePetCareSensorEntityDescription, ...] = (
    SurePetCareSensorEntityDescription(
        key="battery_level",
        native_unit_of_measurement=PERCENTAGE,
        device_class=SensorDeviceClass.BATTERY,
        state_class=SensorStateClass.MEASUREMENT,
        entity_category=EntityCategory.DIAGNOSTIC,
        field=MethodField(path="battery_level"),
    ),
)

SENSOR_DESCRIPTIONS_DEVICE_INFORMATION: tuple[
    SurePetCareSensorEntityDescription, ...
] = (
    SurePetCareSensorEntityDescription(
        key="entity_information",
        translation_key="entity_information",
        icon="mdi:information",
        field=MethodField(
            path=NAME,
            path_extra={
                "household_id": "household_id",
                PRODUCT_ID: PRODUCT_ID,
                "id": "id",
                "parent_device_id": "entity_info.parent_device_id",
            },
        ),
    ),
)

SENSOR_DESCRIPTIONS_RSSI: tuple[SurePetCareSensorEntityDescription, ...] = (
    SurePetCareSensorEntityDescription(
        key="rssi",
        translation_key="rssi",
        native_unit_of_measurement="dBm",
        field=MethodField(
            get_fn=lambda device, r: device.status.signal.device_rssi,
        ),
        device_class=SensorDeviceClass.SIGNAL_STRENGTH,
        entity_category=EntityCategory.DIAGNOSTIC,
    ),
)

SENSOR_DESCRIPTIONS_PET_INFORMATION: tuple[SurePetCareSensorEntityDescription, ...] = (
    SurePetCareSensorEntityDescription(
        key="entity_information",
        translation_key="entity_information",
        icon="mdi:information",
        field=MethodField(
            path=NAME,
            path_extra={
                "household_id": "household_id",
                PRODUCT_ID: PRODUCT_ID,
                "tag": "tag",
                "id": "id",
                "parent_device_id": "entity_info.parent_device_id",
            },
        ),
    ),
)

SENSORS: dict[str, tuple[SurePetCareSensorEntityDescription, ...]] = {
    ProductId.FEEDER_CONNECT: (
        SurePetCareSensorEntityDescription(
            key="bowl_0_weight",
            translation_key="bowl_weight",
            translation_placeholders={"bowl": "One"},
            state_class=SensorStateClass.MEASUREMENT,
            device_class=SensorDeviceClass.WEIGHT,
            native_unit_of_measurement=UnitOfMass.GRAMS,
            field=MethodField(
                get_fn=lambda device, r: index_attr(
                    device.status.bowl_status, 0, "current_weight"
                ),
                get_extra_fn=lambda device, r: {
                    "position": device.status.bowl_status[0].position.name,
                    "food_type": device.control.bowls.settings[0].food_type.name,
                    "last_filled_at": device.status.bowl_status[0].last_filled_at,
                    "last_zeroed_at": device.status.bowl_status[0].last_zeroed_at,
                    "last_fill_weight": device.status.bowl_status[0].last_fill_weight,
                },
            ),
        ),
        SurePetCareSensorEntityDescription(
            key="bowl_1_weight",
            translation_key="bowl_weight",
            translation_placeholders={"bowl": "Two"},
            state_class=SensorStateClass.MEASUREMENT,
            device_class=SensorDeviceClass.WEIGHT,
            native_unit_of_measurement=UnitOfMass.GRAMS,
            field=MethodField(
                get_fn=lambda device, r: index_attr(
                    device.status.bowl_status, 1, "current_weight"
                ),
                get_extra_fn=lambda device, r: {
                    "position": device.status.bowl_status[1].position.name,
                    "food_type": device.control.bowls.settings[1].food_type.name,
                    "substance_type": device.status.bowl_status[1].substance_type,
                    "last_filled_at": device.status.bowl_status[1].last_filled_at,
                    "last_zeroed_at": device.status.bowl_status[1].last_zeroed_at,
                    "last_fill_weight": device.status.bowl_status[1].last_fill_weight,
                },
            ),
        ),
        SurePetCareSensorEntityDescription(
            key="fill_percent",
            translation_key="fill_percent",
            state_class=SensorStateClass.MEASUREMENT,
            native_unit_of_measurement=PERCENTAGE,
            field=MethodField(
                get_fn=lambda device, r: sum_attr(
                    getattr(device.status, "bowl_status", []), "fill_percent"
                ),
                get_extra_fn=lambda device, r: {
                    "bowl_0_fill_percent": index_attr(
                        device.status.bowl_status, 0, "fill_percent"
                    ),
                    "bowl_1_fill_percent": index_attr(
                        device.status.bowl_status, 1, "fill_percent"
                    ),
                },
            ),
        ),
        SurePetCareSensorEntityDescription(
            key="weight_capacity",
            translation_key="weight_capacity",
            state_class=SensorStateClass.MEASUREMENT,
            device_class=SensorDeviceClass.WEIGHT,
            native_unit_of_measurement=UnitOfMass.GRAMS,
            field=MethodField(
                get_fn=lambda device, r: sum(
                    w.target
                    for w in getattr(
                        getattr(getattr(device, "control"), "bowls"), "settings", []
                    )
                ),
                get_extra_fn=lambda device, r: {
                    "bowls": [
                        bowl
                        for bowl in getattr(
                            getattr(getattr(device, "control"), "bowls"), "settings", []
                        )
                    ]
                },
            ),
        ),
        *SENSOR_DESCRIPTIONS_RSSI,
        *SENSOR_DESCRIPTIONS_BATTERY,
        *SENSOR_DESCRIPTIONS_DEVICE_INFORMATION,
    ),
    ProductId.DUAL_SCAN_PET_DOOR: (
        *SENSOR_DESCRIPTIONS_RSSI,
        *SENSOR_DESCRIPTIONS_BATTERY,
        *SENSOR_DESCRIPTIONS_DEVICE_INFORMATION,
    ),
    ProductId.DUAL_SCAN_CONNECT: (
        *SENSOR_DESCRIPTIONS_RSSI,
        *SENSOR_DESCRIPTIONS_BATTERY,
        *SENSOR_DESCRIPTIONS_DEVICE_INFORMATION,
    ),
    ProductId.PET_DOOR: (
        *SENSOR_DESCRIPTIONS_RSSI,
        *SENSOR_DESCRIPTIONS_BATTERY,
        *SENSOR_DESCRIPTIONS_DEVICE_INFORMATION,
    ),
    ProductId.POSEIDON_CONNECT: (
        SurePetCareSensorEntityDescription(
            key="bowl_volume",
            translation_key="bowl_volume",
            translation_placeholders={"bowl": ""},
            state_class=SensorStateClass.MEASUREMENT,
            device_class=SensorDeviceClass.VOLUME,
            native_unit_of_measurement=UnitOfVolume.MILLILITERS,
            field=MethodField(
                get_fn=lambda device, r: index_attr(
                    device.status.bowl_status, 0, "current_weight"
                ),
                get_extra_fn=lambda device, r: {
                    "last_filled_at": device.status.bowl_status[0].last_filled_at,
                    "last_zeroed_at": device.status.bowl_status[0].last_zeroed_at,
                    "last_fill_weight": device.status.bowl_status[0].last_fill_weight,
                },
            ),
        ),
        SurePetCareSensorEntityDescription(
            key="fill_percent",
            translation_key="fill_percent",
            state_class=SensorStateClass.MEASUREMENT,
            native_unit_of_measurement=PERCENTAGE,
            field=MethodField(
                get_fn=lambda device, r: sum_attr(
                    getattr(device.status, "bowl_status", []), "fill_percent"
                ),
            ),
        ),
        *SENSOR_DESCRIPTIONS_RSSI,
        *SENSOR_DESCRIPTIONS_BATTERY,
        *SENSOR_DESCRIPTIONS_DEVICE_INFORMATION,
    ),
    ProductId.HUB: (),
    ProductId.PET: (
        SurePetCareSensorEntityDescription(
            key="feeding",
            translation_key="feeding",
            device_class=SensorDeviceClass.WEIGHT,
            state_class=SensorStateClass.MEASUREMENT,
            native_unit_of_measurement=UnitOfMass.GRAMS,
            entity_registry_enabled_default=False,
            field=MethodField(
                get_fn=lambda device, r: abs(change[0] + change[1])
                if (change := getattr(device.status.feeding, "change", []))
                else None,
                get_extra_fn=lambda device, config_options: {
                    "device_id": device.status.feeding.device_id,
                    "id": device.status.feeding.id,
                    "at": device.status.feeding.at,
                    "tag_id": device.status.feeding.tag_id,
                    "change_0": abs(device.status.feeding.change[0]),
                    "change_1": abs(device.status.feeding.change[1]),
                },
            ),
        ),
        SurePetCareSensorEntityDescription(
            key="position",
            translation_key="position",
            entity_registry_enabled_default=False,
            field=MethodField(
                get_fn=get_location,
                path_extra={
                    "device_id": "status.activity.device_id",
                    "id": "status.activity.id",
                    "since": "status.activity.since",
                    "where": "status.activity.where",
                    "tag_id": "status.activity.tag_id",
                },
            ),
        ),
        SurePetCareSensorEntityDescription(
            key="drinking",
            translation_key="drinking",
            device_class=SensorDeviceClass.WEIGHT,
            state_class=SensorStateClass.MEASUREMENT,
            native_unit_of_measurement=UnitOfMass.GRAMS,
            entity_registry_enabled_default=False,
            field=MethodField(
                get_fn=lambda device, r: abs(change[0] + change[1])
                if (change := getattr(device.status.drinking, "change", []))
                else None,
                get_extra_fn=lambda device, config_options: {
                    "device_id": device.status.drinking.device_id,
                    "id": device.status.drinking.id,
                    "at": device.status.drinking.at,
                    "tag_id": device.status.drinking.tag_id,
                    "change_0": abs(device.status.drinking.change[0]),
                    "change_1": abs(device.status.drinking.change[1]),
                },
            ),
        ),
        SurePetCareSensorEntityDescription(
            key="devices",
            translation_key="devices",
            native_unit_of_measurement="pcs",
            field=MethodField(
                get_fn=lambda device, r: len(
                    getattr(device.status, "devices", []) or []
                ),
                get_extra_fn=lambda device, r: {
                    "devices": [
                        d.id for d in getattr(device.status, "devices", []) or []
                    ]
                },
            ),
        ),
        SurePetCareSensorEntityDescription(
            key="last_activity",
            translation_key="last_activity",
            entity_registry_enabled_default=False,
            field=MethodField(
                get_fn=lambda device, r: option_name(
                    r, (device.last_activity() or [None, None])[1]
                )
                if device.last_activity()
                else None,
            ),
        ),
        *SENSOR_DESCRIPTIONS_PET_INFORMATION,
    ),
}


async def async_setup_entry(
    hass: HomeAssistant,
    config_entry: ConfigEntry,
    async_add_entities: AddConfigEntryEntitiesCallback,
) -> None:
    """Set up SurePetCare sensors for each matching device."""
    coordinator_data = hass.data[DOMAIN][config_entry.entry_id][COORDINATOR]
    client = coordinator_data[KEY_API]

    entities = []
    for device_coordinator in coordinator_data[COORDINATOR_DICT].values():
        descriptions = SENSORS.get(device_coordinator.product_id, ())
        entities.extend(
            [
                SurePetCareSensor(
                    device_coordinator,
                    client,
                    description=description,
                )
                for description in descriptions
                if should_add_entity(
                    description, device_coordinator.data, config_entry.options
                )
            ]
        )
    async_add_entities(entities, update_before_add=True)


class SurePetCareSensor(SurePetCareBaseEntity, SensorEntity):
    """The platform class required by Home Assistant."""

    entity_description: SurePetCareSensorEntityDescription

    def __init__(
        self,
        device_coordinator: SurePetCareDeviceDataUpdateCoordinator,
        client: SurePetcareClient,
        description: SurePetCareSensorEntityDescription,
    ) -> None:
        """Initialize a Surepetcare sensor."""
        super().__init__(
            device_coordinator=device_coordinator,
            client=client,
        )
        self.entity_description = description
        self._attr_unique_id = f"{self._attr_unique_id}-{description.key}"

    @property
    def entity_picture(self) -> str | None:
        """Return the entity picture URL."""
        if self.entity_description.icon:
            return self.coordinator.data.photo
        return None
