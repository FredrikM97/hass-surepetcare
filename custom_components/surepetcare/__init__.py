"""TODO."""

import logging
from typing import Any, List

from surepetcare.client import SurePetcareClient
from surepetcare.household import Household
from surepetcare.enums import ProductId

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import EVENT_HOMEASSISTANT_STOP, Platform
from homeassistant.core import Event, HomeAssistant, callback
from homeassistant.helpers import device_registry as dr

from .const import (
    CLIENT_DEVICE_ID,
    COORDINATOR,
    COORDINATOR_DICT,
    DOMAIN,
    FACTORY,
    KEY_API,
    TOKEN,
)
from .coordinator import SurePetCareDeviceDataUpdateCoordinator

logger = logging.getLogger(__name__)

PLATFORMS = [Platform.BINARY_SENSOR, Platform.SENSOR]


async def async_setup_entry(
    hass: HomeAssistant,
    entry: ConfigEntry,
) -> bool:
    """TODO."""
    logger.info("async_setup_entry called for entry_id=%s", entry.entry_id)

    surepetcare_data: dict[str, Any] = {}
    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = surepetcare_data

    client: SurePetcareClient = SurePetcareClient()
    # Try login with token+device_id, else fallback to email+password
    token = entry.data.get(TOKEN)
    device_id = entry.data.get(CLIENT_DEVICE_ID)
    email = entry.data.get("email")
    password = entry.data.get("password")
    if token and device_id:
        await client.login(token=token, device_id=device_id)
    elif email and password:
        await client.login(email=email, password=password)
    else:
        raise Exception("No valid credentials found in config entry data")
    surepetcare_data[FACTORY] = client

    async def on_hass_stop(event: Event) -> None:
        """Close connection when hass stops."""
        await client.close()

    # Setup listeners
    entry.async_on_unload(
        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, on_hass_stop)
    )
    try:
        households: List[Household] = await client.api(Household.get_households())
        entities = []
        for household in households:
            entities.extend(await client.api(household.get_pets()))
            entities.extend(await client.api(household.get_devices()))
            # Add each hub device to hubs list
        await client.close()
    except Exception as exc:
        await client.close()
        raise Exception("Configuration not finished") from exc

    remove_stale_devices(hass, entry, entities)
    # Setup the device coordinators
    coordinator_data = {
        KEY_API: client,
        COORDINATOR_DICT: {},
    }

    for device in entities:
        if device.product != ProductId.HUB:
            continue
        entities.remove(device)
        device_id = str(device.id)
        if device_id not in coordinator_data[COORDINATOR_DICT]:
            coordinator = SurePetCareDeviceDataUpdateCoordinator(
                hass=hass,
                config_entry=entry,
                client=client,
                device=device,
            )
            await coordinator.async_config_entry_first_refresh()

            coordinator_data[COORDINATOR_DICT][device_id] = coordinator
        else:
            logger.warning("Coordinator already exists for device %s", device_id)

    for device in entities:
        device_id = str(device.id)
        if device_id not in coordinator_data[COORDINATOR_DICT]:
            coordinator = SurePetCareDeviceDataUpdateCoordinator(
                hass=hass,
                config_entry=entry,
                client=client,
                device=device,
            )
            await coordinator.async_config_entry_first_refresh()

            coordinator_data[COORDINATOR_DICT][device_id] = coordinator
        else:
            logger.warning("Coordinator already exists for device %s", device_id)

    surepetcare_data[COORDINATOR] = coordinator_data

    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)

    return True


async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Unload a config entry."""
    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
    if unload_ok:
        surepetcare_data = hass.data[DOMAIN].pop(entry.entry_id)
        client: SurePetcareClient = surepetcare_data[FACTORY]
        await client.close()

    return unload_ok


@callback
def remove_stale_devices(
    hass: HomeAssistant, config_entry: ConfigEntry, devices: list[Any]
) -> None:
    """Remove stale devices from device registry. TODO: Work in progress and not functional yet"""

    device_registry = dr.async_get(hass)
    device_entries = dr.async_entries_for_config_entry(
        device_registry, config_entry.entry_id
    )
    all_device_ids = {str(device.id) for device in devices}
    for device_entry in device_entries:
        device_id: str | None = None  # Only define here
        # Check that device part of DOMAIN
        for identifier in device_entry.identifiers:
            if identifier[0] != DOMAIN:
                continue

            _id = identifier[1]
            device_id = str(_id)
        if device_id is None or device_id not in all_device_ids:
            logger.info(
                "Removing stale device entry %s for config entry %s",
                device_entry.id,
                config_entry.entry_id,
            )
            device_registry.async_update_device(
                device_entry.id, remove_config_entry_id=config_entry.entry_id
            )
