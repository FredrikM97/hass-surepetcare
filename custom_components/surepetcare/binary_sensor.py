"""Support for Sure Petcare binary sensors."""

from dataclasses import dataclass
from datetime import datetime
import logging
from typing import cast

from surepcio.enums import ProductId
from surepcio.devices.device import SurePetCareBase
from surepcio.devices.dual_scan_connect import Curfew
from homeassistant.components.binary_sensor import (
    BinarySensorDeviceClass,
    BinarySensorEntity,
    BinarySensorEntityDescription,
)
from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.helpers.entity import EntityCategory
from homeassistant.helpers.entity_platform import AddEntitiesCallback

from .const import COORDINATOR, COORDINATOR_DICT, DOMAIN, KEY_API
from .coordinator import SurePetCareDeviceDataUpdateCoordinator
from .entity import (
    SurePetCareBaseEntity,
    SurePetCareBaseEntityDescription,
)

logger = logging.getLogger(__name__)


def _next_enabled_future_curfew(device: SurePetCareBase, r: ConfigEntry) -> bool | None:
    curfews: list[Curfew] | Curfew = device.control.curfew

    if curfews is None:
        return None

    if not isinstance(curfews, list):
        curfews = [curfews]

    now = datetime.now().time()
    return any(c.enabled and c.lock_time <= now <= c.unlock_time for c in curfews)


@dataclass(frozen=True, kw_only=True)
class SurePetCareBinarySensorEntityDescription(
    SurePetCareBaseEntityDescription, BinarySensorEntityDescription
):
    """Describes SurePetCare.binary_sensor entity."""


SENSOR_DESCRIPTIONS_AVAILABLE: tuple[SurePetCareBinarySensorEntityDescription, ...] = (
    SurePetCareBinarySensorEntityDescription(
        key="connectivity",
        translation_key="connectivity",
        device_class=BinarySensorDeviceClass.CONNECTIVITY,
        field="available",
    ),
)

SENSORS: dict[str, tuple[SurePetCareBinarySensorEntityDescription, ...]] = {
    ProductId.FEEDER_CONNECT: (
        SurePetCareBinarySensorEntityDescription(
            key="learn_mode",
            translation_key="learn_mode",
            field="status.learn_mode",
            entity_category=EntityCategory.DIAGNOSTIC,
            entity_registry_enabled_default=False,
        ),
        *SENSOR_DESCRIPTIONS_AVAILABLE,
    ),
    ProductId.DUAL_SCAN_CONNECT: (
        SurePetCareBinarySensorEntityDescription(
            key="learn_mode",
            translation_key="learn_mode",
            field="status.learn_mode",
            entity_category=EntityCategory.DIAGNOSTIC,
            entity_registry_enabled_default=False,
        ),
        SurePetCareBinarySensorEntityDescription(
            key="curfew",
            translation_key="curfew_active",
            field_fn=_next_enabled_future_curfew,
            extra_field="control.curfew",
        ),
        *SENSOR_DESCRIPTIONS_AVAILABLE,
    ),
    ProductId.PET_DOOR: (
        SurePetCareBinarySensorEntityDescription(
            key="learn_mode",
            translation_key="learn_mode",
            field="status.learn_mode",
            entity_category=EntityCategory.DIAGNOSTIC,
            entity_registry_enabled_default=False,
        ),
        SurePetCareBinarySensorEntityDescription(
            key="curfew",
            translation_key="curfew_active",
            field_fn=_next_enabled_future_curfew,
            extra_field="control.curfew",
        ),
    ),
    ProductId.DUAL_SCAN_PET_DOOR: (*SENSOR_DESCRIPTIONS_AVAILABLE,),
    ProductId.HUB: (*SENSOR_DESCRIPTIONS_AVAILABLE,),
}


async def async_setup_entry(
    hass: HomeAssistant,
    config_entry: ConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
    """Set up a Surepetcare config entry."""
    coordinator_data = hass.data[DOMAIN][config_entry.entry_id][COORDINATOR]
    client = coordinator_data[KEY_API]

    entities = []
    for device_coordinator in coordinator_data[COORDINATOR_DICT].values():
        descriptions = SENSORS.get(device_coordinator.product_id, ())
        entities.extend(
            [
                SurePetCareBinarySensor(
                    device_coordinator,
                    client,
                    description=description,
                )
                for description in descriptions
            ]
        )
    async_add_entities(entities, update_before_add=True)


class SurePetCareBinarySensor(SurePetCareBaseEntity, BinarySensorEntity):
    """The platform class required by Home Assistant."""

    entity_description: SurePetCareBinarySensorEntityDescription

    def __init__(
        self,
        device_coordinator: SurePetCareDeviceDataUpdateCoordinator,
        client,
        description: SurePetCareBinarySensorEntityDescription,
    ) -> None:
        """Initialize a SurePetCare binary sensor."""
        super().__init__(
            device_coordinator=device_coordinator,
            client=client,
        )

        self.entity_description = description

        self._attr_unique_id = f"{self._attr_unique_id}-{description.key}"

    @property
    def is_on(self) -> bool | None:
        """Return true if the binary sensor is on."""
        return cast(bool, self._convert_value())
