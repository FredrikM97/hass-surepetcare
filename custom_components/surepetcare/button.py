"""Support for Sure Petcare Button."""

from dataclasses import dataclass
import logging
from typing import Any
from surepcio.enums import ProductId, HubPairMode
from surepcio import SurePetcareClient
from homeassistant.components.button import ButtonEntity, ButtonEntityDescription
from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.helpers.entity_platform import AddConfigEntryEntitiesCallback
from homeassistant.helpers.entity import EntityCategory
from custom_components.surepetcare.helper import MethodField, should_add_entity


from .const import (
    COORDINATOR,
    COORDINATOR_DICT,
    DOMAIN,
    KEY_API,
)
from .coordinator import SurePetCareDeviceDataUpdateCoordinator
from .entity import (
    SurePetCareBaseEntity,
    SurePetCareBaseEntityDescription,
)

logger = logging.getLogger(__name__)

@dataclass(frozen=True, slots=True)
class ButtonMethodField(MethodField):
    """MethodField for button-like entities, supporting on mapping."""

    on: Any = True
    def set(self, device: object, config: dict, value: Any) -> Any:
        if value is True and self.on:
            value = self.on
        return MethodField.set(self, device, config, value)



@dataclass(frozen=True, kw_only=True)
class SurePetCareButtonEntityDescription(
    SurePetCareBaseEntityDescription, ButtonEntityDescription
):
    """Describes SurePetCare button entity."""


BUTTONS: dict[str, tuple[SurePetCareButtonEntityDescription, ...]] = {    
    ProductId.HUB: (
         SurePetCareButtonEntityDescription(
            key="pairing_mode",
            translation_key="pairing_mode",
            field=ButtonMethodField(
                path="control.pairing_mode", on=HubPairMode.ON
            ),
            entity_category=EntityCategory.CONFIG,
        ),
    )
    
}


async def async_setup_entry(
    hass: HomeAssistant,
    config_entry: ConfigEntry,
    async_add_entities: AddConfigEntryEntitiesCallback,
) -> None:
    """Set up SurePetCare sensors for each matching device."""
    coordinator_data = hass.data[DOMAIN][config_entry.entry_id][COORDINATOR]
    client = coordinator_data[KEY_API]

    entities = []
    for device_coordinator in coordinator_data[COORDINATOR_DICT].values():
        descriptions = BUTTONS.get(device_coordinator.product_id, ())
        entities.extend(
            [
                SurePetCareButton(
                    device_coordinator,
                    client,
                    description=description,
                )
                for description in descriptions
                if should_add_entity(
                    description, device_coordinator.data, config_entry.options
                )
            ]
        )
    async_add_entities(entities, update_before_add=True)


class SurePetCareButton(SurePetCareBaseEntity, ButtonEntity):
    """The platform class required by Home Assistant."""

    entity_description: SurePetCareButtonEntityDescription

    def __init__(
        self,
        device_coordinator: SurePetCareDeviceDataUpdateCoordinator,
        client: SurePetcareClient,
        description: SurePetCareButtonEntityDescription,
    ) -> None:
        """Initialize a Surepetcare sensor."""
        super().__init__(
            device_coordinator=device_coordinator,
            client=client,
        )
        self.entity_description = description
        self._attr_unique_id = f"{self._attr_unique_id}-{description.key}"

    async def async_press(self) -> None:
        """Press the button."""
        await self.send_command(1)
