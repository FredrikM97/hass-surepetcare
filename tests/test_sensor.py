import json
from pathlib import Path
import pytest
from syrupy.assertion import SnapshotAssertion
from unittest.mock import MagicMock
from custom_components.surepetcare.const import DOMAIN
from custom_components.surepetcare.sensor import (
    async_setup_entry,
)
from tests.conftest import (
    FIXTURES,
    create_device_from_fixture,
    extract_sensor_outputs,
    setup_coordinator,
)

from pytest_homeassistant_custom_component.common import MockConfigEntry


@pytest.mark.parametrize("fixture_file", FIXTURES)
@pytest.mark.asyncio
async def test_sensor_snapshot_from_fixture(
    hass, snapshot: SnapshotAssertion, fixture_file
):
    fixture_data = json.loads((Path("tests/fixtures") / fixture_file).read_text())
    config_entry = MockConfigEntry(
        domain=DOMAIN, data={"token": "abc", "device_id": "123"}
    )
    config_entry.add_to_hass(hass)

    device = create_device_from_fixture(fixture_data)
    setup_coordinator(hass, config_entry, device)

    async_add_entities = MagicMock()
    await async_setup_entry(hass, config_entry, async_add_entities)

    assert extract_sensor_outputs(async_add_entities) == snapshot
