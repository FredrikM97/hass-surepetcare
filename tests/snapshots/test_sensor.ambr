# serializer version: 1
# name: test_sensor_snapshot_from_fixture[feeder_connect.json]
  dict({
    '269654-battery_level': dict({
      'device_class': <SensorDeviceClass.BATTERY: 'battery'>,
      'extra_attributes': None,
      'unit': '%',
      'value': 13,
    }),
    '269654-bowl_1_weight': dict({
      'device_class': <SensorDeviceClass.WEIGHT: 'weight'>,
      'extra_attributes': dict({
        'current_weight': 32.16,
        'food_type': 1,
        'last_fill_weight': 32.16,
        'last_filled_at': '2025-07-07T10:29:29+00:00',
        'last_zeroed_at': '2025-05-23T10:25:29+00:00',
        'position': 0,
        'substance_type': 2,
      }),
      'unit': <UnitOfMass.GRAMS: 'g'>,
      'value': 32.16,
    }),
    '269654-bowl_2_weight': dict({
      'device_class': <SensorDeviceClass.WEIGHT: 'weight'>,
      'extra_attributes': dict({
        'current_weight': -27.2,
        'food_type': 1,
        'last_fill_weight': -27.2,
        'last_filled_at': '2025-07-07T10:29:29+00:00',
        'last_zeroed_at': '2025-05-23T10:25:29+00:00',
        'position': 1,
        'substance_type': 2,
      }),
      'unit': <UnitOfMass.GRAMS: 'g'>,
      'value': -27.2,
    }),
    '269654-entity_information': dict({
      'device_class': None,
      'extra_attributes': dict({
        'household_id': 222527,
        'id': 269654,
        'parent_device_id': 295972,
        'product_id': 4,
      }),
      'unit': None,
      'value': 'Maui matl√•da (old) ',
    }),
    '269654-lid_delay': dict({
      'device_class': None,
      'extra_attributes': None,
      'unit': None,
      'value': None,
    }),
    '269654-rssi': dict({
      'device_class': None,
      'extra_attributes': None,
      'unit': None,
      'value': -81,
    }),
    '269654-tare': dict({
      'device_class': None,
      'extra_attributes': None,
      'unit': None,
      'value': 2,
    }),
    '269654-training_mode': dict({
      'device_class': None,
      'extra_attributes': None,
      'unit': None,
      'value': 0,
    }),
    '269654-weight_capacity': dict({
      'device_class': None,
      'extra_attributes': dict({
        'food_type': 1,
      }),
      'unit': None,
      'value': 0,
    }),
  })
# ---
# name: test_sensor_snapshot_from_fixture[hub.json]
  dict({
  })
# ---
# name: test_sensor_snapshot_from_fixture[pet.json]
  dict({
    '472721-entity_information': dict({
      'device_class': None,
      'extra_attributes': dict({
        'household_id': 222527,
        'id': 472721,
        'parent_device_id': None,
        'product_id': 0,
        'tag': 60978,
      }),
      'unit': None,
      'value': 'Maui',
    }),
    '472721-feeding': dict({
      'device_class': <SensorDeviceClass.WEIGHT: 'weight'>,
      'extra_attributes': None,
      'unit': <UnitOfMass.GRAMS: 'g'>,
      'value': 0,
    }),
  })
# ---
